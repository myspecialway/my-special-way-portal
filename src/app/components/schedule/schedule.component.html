<table class="mat-table schedule-table" role="grid">
  <thead>
    <tr class="mat-header-row" role="row">
      <th class="mat-header-cell cdk-column-position mat-column-position" role="columnheader" colspan="2">שעה</th>
      <th class="mat-header-cell cdk-column-position mat-column-position" role="columnheader" *ngFor="let day of daysLabels">{{day}}</th>
    </tr>
  </thead>
  <tbody>
    <tr class="mat-row" role="row" *ngFor="let row of schedule; index as hourIndex">
      <td class="mat-cell cdk-column-position mat-column-position" role="gridcell">{{hourIndex + 1}}</td>
      <td class="mat-cell cdk-column-position mat-column-position" role="gridcell">{{hoursLabels[hourIndex]}}</td>
      <td 
        class="mat-cell cdk-column-position mat-column-position"
        *ngFor="let timeslot of row; index as dayIndex" 
        role="gridcell" 
        (click)="timeSlotClicked.emit({hourIndex: hourIndex, dayIndex: dayIndex})"
      >
        <div *ngIf="timeslot.lesson" class="ts-full">
          <span class="ts-lesson">{{timeslot.lesson.title}}</span>
          <span class="ts-location" *ngIf="timeslot.location">{{timeslot.location.name}}</span>
        </div>
        <div *ngIf="!timeslot.lesson" class="ts-empty">
          <i class="material-icons" >
            add
          </i>
        </div>
      </td>
    </tr>
  </tbody>
</table>
