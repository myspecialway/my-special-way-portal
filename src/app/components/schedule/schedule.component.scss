$hover: rgba(106, 0, 207, 0.2);

.schedule-table {
    width: 100%;
    direction: rtl;
    font-family: "Rubik", "Helvetica", "Arial", sans-serif;
    color: #222;
    th.mat-header-cell {
        background-color: #fafafa;
        font-size: 16px;
        font-weight: bold;
        color: #222;
        border: 0;
    }
    td.mat-cell, td.mat-footer-cell, th.mat-header-cell {
        text-align: right;
        padding: 20px 14px;
    }
    td.mat-cell, td.mat-footer-cell {
        cursor: pointer;
        border: 1px solid #e0e0e0;
        font-size: 18px;
        &:hover {
            background-color: $hover;
        }
    }

    td.mat-cell:first-child, td:nth-child(2) {
        text-align: center;
        direction: initial;
        font-weight: 400;
    }
    
    .ts-lesson, .ts-location {
        font-family: Rubik;
        color: #222;
        display: block;
    }
    .ts-lesson {
        font-size: 18px;
        font-weight: 400;
    }
    .ts-location {
        font-size: 14px;
    }
    .ts-empty {
        text-align: center;
        color: #6a00cf;
        opacity: .5;
        &:hover {
            opacity: 1;
        }
    }
}
