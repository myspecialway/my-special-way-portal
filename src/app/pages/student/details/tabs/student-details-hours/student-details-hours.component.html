<div class="msw-student-details-hours">
  <div class="msw-student-header-nav-name" *ngIf="!isNewStudent && student">
    {{ student.firstname }} {{ student.lastname }}
  </div>
    <app-schedule [daysLabels]="scheduleService.daysLabels" [hoursLabels]="scheduleService.hoursLabels" [schedule]="schedule" (timeSlotClicked)="onTimeSlotClick($event)"></app-schedule>
</div>
