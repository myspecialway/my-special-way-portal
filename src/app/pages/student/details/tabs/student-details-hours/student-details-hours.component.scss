.msw-student-header-nav-name {
  font-size: 32px;
  padding-top: 5px;
  letter-spacing: 0.6px;
  text-align: right;
  color: #222222;
  position: absolute;
  top: 125px;
  right: 38px;
}
