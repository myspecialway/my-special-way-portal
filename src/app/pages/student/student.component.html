<div data-test-id="students-page" class="main-content msw-student-container">
  <div class="container-fluid">
    <div class="header-menu-container msw-table-tools-header">
      <div class="header-table-count">
        <div>
          <span class="header-table-count-label">תלמידים</span>
          <span class="header-table-count-number">({{dataSource.data.length}})</span>
        </div>
      </div>
      <div class="example-header msw-table-tools-header">
        <button data-test-id="add-new-student-button" mat-raised-button color="primary" routerLink="/student/_new_"  *ngIf="mayAddStudent">
          <i class="material-icons">
            add
          </i>
          הוסף תלמיד חדש
        </button>
      </div>
    </div>
    <div class="row">
      <div class="col-lg-12 col-md-12 col-sm-12">
        <mat-table [dataSource]="dataSource" class="msw-table" matSort matSortDisableClear>
          <ng-container matColumnDef="studentName">
            <mat-header-cell *matHeaderCellDef>שם</mat-header-cell>
            <mat-cell *matCellDef="let row" [routerLink]="['/student', row._id]">{{ row.firstname }} {{ row.lastname }}</mat-cell>
          </ng-container>

          <ng-container matColumnDef="gradeId">
            <mat-header-cell *matHeaderCellDef>כיתה</mat-header-cell>
            <mat-cell *matCellDef="let row" [routerLink]="['/student', row._id]">{{(row.class && row.class.name) || '' }}</mat-cell>
          </ng-container>

          <ng-container matColumnDef="username">
            <mat-header-cell *matHeaderCellDef mat-sort-header>שם משתמש</mat-header-cell>
            <mat-cell *matCellDef="let row" class="username" [routerLink]="['/student', row._id]">{{ row.username }}</mat-cell>
          </ng-container>

          <ng-container matColumnDef="deleteUser">
            <mat-header-cell *matHeaderCellDef></mat-header-cell>
            <mat-cell *matCellDef="let row">
              <button mat-icon-button color="primary" [attr.data-test-id]="'delete-user-button-' + row.username" [disabled]="!mayDeleteStudent"
                      (click)="deleteStudent(row._id, row.firstname, row.lastname, row.gradeId, row.gender)"><i class="material-icons"> delete </i> </button>
            </mat-cell>
          </ng-container>

          <mat-header-row *matHeaderRowDef="displayedColumns"></mat-header-row>
          <mat-row *matRowDef="let row; columns: displayedColumns;"></mat-row>
        </mat-table>
      </div>
    </div>
  </div>
</div>
