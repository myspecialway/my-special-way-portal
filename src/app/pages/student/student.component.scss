@import '../../../assets/scss/msw-table';

.header-table-count {
  width: 164px;
  height: 38px;
  font-size: 32px;
  font-weight: normal;
  font-style: normal;
  font-stretch: normal;
  line-height: normal;
  letter-spacing: 0.6px;
  text-align: right;
  color: #222222;
}
.header-table-count-number {
  font-size: 18px;
}

.header-table-count-label {
  padding-left: 10px;
}

.header-menu-container {
  padding-bottom: 29px;
}

.msw-student-container .mat-cell:last-child,
.msw-student-container .mat-header-cell:last-child {
  flex-grow: 0;
  flex-basis: 100px;
}
.msw-student-container .mat-header-cell {
  font-weight: bold;
}

.mat-column-deleteUser button {
  display: none;
}
.mat-row:hover {
  .mat-cell {
    color: #8c37cd;
  }
  .mat-column-deleteUser button {
    display: block;
  }
}
