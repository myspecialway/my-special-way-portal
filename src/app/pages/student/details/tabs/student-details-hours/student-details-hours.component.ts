import { Component, OnInit } from '@angular/core';
import { ActivatedRoute } from '@angular/router';
import { ScheduleService } from '../../../../../services/schedule/schedule.service';
import { StudentService } from '../../../services/student.service';
import { TimeSlot } from '../../../../../models/timeslot.model';
import Student, { StudentQuery } from '../../../../../models/student.model';
import { TimeSlotIndexes } from '../../../../../components/schedule/schedule.component';
import { ScheduleDialogData } from '../../../../../components/schedule/schedule-dialog/schedule-dialog-data.model';
import { MatDialog } from '@angular/material';
import { ScheduleDialogComponent } from '../../../../../components/schedule/schedule-dialog/schedule.dialog';
import { SubscriptionCleaner } from '../../../../../decorators/SubscriptionCleaner.decorator';
import { Class } from '../../../../../models/class.model';

@Component({
  selector: 'app-student-details-hours',
  templateUrl: './student-details-hours.component.html',
  styleUrls: ['./student-details-hours.component.scss'],
})
export class StudentDetailsHoursComponent implements OnInit {
  sub: any;
  id: string;
  schedule: TimeSlot[][];
  student: Student;

  @SubscriptionCleaner()
  subCollector;

  constructor(
    private route: ActivatedRoute,
    public scheduleService: ScheduleService,
    private studentService: StudentService,
    private dialog: MatDialog,
  ) {}

  async ngOnInit() {
    if (!this.route.parent) {
      return;
    }
    this.id = this.route.parent.snapshot.params.idOrNew;
    try {
      this.student = await this.studentService.getById(this.id);
      if (!this.student.class) {
        this.student.class = new Class();
      }
      this.initSchedule();
    } catch (err) {
      throw err;
    }
  }

  private initSchedule() {
    const schedule = this.student.schedule || [];
    this.schedule = this.scheduleService.buildScheduleFromTimeslots(
      this.scheduleService.hoursLabels.length,
      this.scheduleService.daysLabels.length,
      schedule,
    );
  }

  onTimeSlotClick(indexes: TimeSlotIndexes) {
    const dialogData = this.getDialogData(indexes);

    const dialogRef = this.dialog.open(ScheduleDialogComponent, {
      data: dialogData,
      height: '375px',
      width: '320px',
    });

    this.subCollector.add(
      dialogRef.afterClosed().subscribe(async (data: ScheduleDialogData) => {
        if (!data) {
          return;
        }
        const onlyCustomizedSlots: TimeSlot[] = this.student.schedule.filter((slot) => slot.customized);
        const newCustomizedSlot: TimeSlot = {
          index: data.index,
          lesson: data.lesson,
          location: data.location,
          customized: true,
        };
        const newCustomizedSchedule = [...onlyCustomizedSlots, newCustomizedSlot];

        const tempStudent: StudentQuery = {
          _id: this.student._id,
          username: this.student.username,
          firstname: this.student.firstname,
          lastname: this.student.lastname,
          gender: this.student.gender,
          password: this.student.password,
          class_id: this.student.class._id,
          schedule: newCustomizedSchedule,
        };
        try {
          await this.studentService.update(tempStudent);
          this.student = await this.studentService.getById(this.id);
          this.initSchedule();
        } catch (error) {
          console.log(error);
        }
      }),
    );
  }

  getDialogData(indexes: TimeSlotIndexes) {
    const { hourIndex, dayIndex } = indexes;

    return {
      index: `${hourIndex}_${dayIndex}`,
      lesson: this.schedule[hourIndex][dayIndex].lesson,
      location: this.schedule[hourIndex][dayIndex].location,
      hour: this.scheduleService.hoursLabels[hourIndex],
      day: this.scheduleService.daysLabels[dayIndex],
    } as ScheduleDialogData;
  }
}
