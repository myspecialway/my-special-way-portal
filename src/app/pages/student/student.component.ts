import { Component, OnInit, ViewChild, ElementRef } from '@angular/core';
import { MatTableDataSource, MatSort, MatDialog } from '@angular/material';
import { DeleteStudentDialogComponent } from './dialogs/delete/delete-student.dialog';
import { StudentService } from './services/student.service';
import Student from '../../models/student.model';
@Component({
  selector: 'app-student',
  templateUrl: './student.component.html',
  styleUrls: ['./student.component.scss'],
})
export class StudentComponent implements OnInit {
  displayedColumns = ['studentName', 'gradeId', 'username', 'deleteUser'];
  dataSource = new MatTableDataSource<Student>();

  @ViewChild(MatSort)
  sort: MatSort;
  @ViewChild('table')
  table: ElementRef;

  constructor(private studentService: StudentService, private dialog: MatDialog) {}

  async ngOnInit() {
    this.dataSource.sort = this.sort;
    this.studentService.getAllStudents().subscribe((data) => {
      this.dataSource.data = [...data];
    });
  }

  async deleteStudent(id: number, firstName: string, lastName: string, gradeId: string) {
    const dialogRef = this.dialog.open(DeleteStudentDialogComponent, {
      data: { id, firstName, lastName, gradeId },
    });

    dialogRef.afterClosed().subscribe(async (deletionConfirmed) => {
      if (!deletionConfirmed) {
        return;
      }

      try {
        await this.studentService.delete(id);
      } catch (error) {
        // TODO: implement error handling on UI
        console.error('Error handling not implemented');
        throw error;
      }
    });
  }
}
