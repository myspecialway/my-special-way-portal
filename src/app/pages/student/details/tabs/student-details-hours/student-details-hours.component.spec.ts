import { Subscription } from 'rxjs/Subscription';
import { scheduleTestData } from './../../../../../../mocks/assets/schedule.mock';
import { ScheduleDialogComponent } from './../../../../../components/schedule/schedule-dialog/schedule.dialog';
import { ScheduleDialogData } from './../../../../../components/schedule/schedule-dialog/schedule-dialog-data.model';
jest.mock('../../../services/student.service');

import { TestBed, ComponentFixture } from '@angular/core/testing';
import { NO_ERRORS_SCHEMA } from '@angular/core';
import { Platform } from '@angular/cdk/platform';
import { Observable, Subject } from 'rxjs-compat';
import { ActivatedRoute, Router, RouterModule } from '@angular/router';
import { StudentDetailsHoursComponent } from './student-details-hours.component';
import { StudentDetailsComponent } from '../../student-details.component';
import { StudentService } from '../../../services/student.service';
import { ScheduleService } from '../../../../../services/schedule/schedule.service';
import { MatDialog } from '@angular/material';
import { TranslateService, TranslateLoader, TranslateModule } from '@ngx-translate/core';
import { TranslateCustomLoader } from '../../../../../../mocks/translate.stub';

describe('Student Details Hours Component', () => {
  const sub = new Subscription();
  let scheduleDialogMock: Partial<MatDialog>;
  let component: StudentDetailsHoursComponent;
  let fixture: ComponentFixture<StudentDetailsHoursComponent>;
  let observableAfterClosed: Subject<ScheduleDialogData>;
  let afterClosedMockFn: jest.Mock;
  let studentServiceMock: Partial<StudentService>;
  let mockedScheduleDialogData;
  const mockedIndexes = { hourIndex: 0, dayIndex: 0 };

  const mockedClass = {
    _id: '5b217b030825622c97d3757f',
    grade: 'a',
    name: 'טיטאן',
    schedule: [
      {
        index: '1_0',
        lesson: {
          _id: '5b2abc74572e7619a628c11c',
          title: 'test lesson',
          icon: 'test-icon',
        },
        location: {
          _id: '5b5596a7739a882933edd4fc',
          disabled: false,
          name: 'test location',
          position: {
            latitude: 0,
            longitude: 0,
            floor: 1,
          },
        },
      },
    ],
  };
  beforeEach(() => {
    setup();
  });

  afterEach(() => {
    sub.unsubscribe();
  });

  describe('with _new_ student path', () => {
    it('should render the component as described in snapshot', () => {
      expect(fixture).toMatchSnapshot();
    });
  });

  describe('with any student path', async () => {
    it('should trigger dialog open when onTimeSlotClick', () => {
      const schedule = mockedClass.schedule;
      component.schedule = component.scheduleService.buildScheduleFromTimeslots(
        component.scheduleService.hoursLabels.length,
        component.scheduleService.daysLabels.length,
        schedule,
      );

      component.onTimeSlotClick(mockedIndexes);
      expect(scheduleDialogMock.open).toBeCalled();
    });

    it('should trigger dialog open with dialog data when onTimeSlotClick', () => {
      const schedule = mockedClass.schedule;
      component.schedule = component.scheduleService.buildScheduleFromTimeslots(
        component.scheduleService.hoursLabels.length,
        component.scheduleService.daysLabels.length,
        schedule,
      );
      component.onTimeSlotClick(mockedIndexes);
      expect(scheduleDialogMock.open).toBeCalledWith(ScheduleDialogComponent, {
        data: component.getDialogData(mockedIndexes),
        height: '375px',
        width: '320px',
      });
    });
  });

  it('should render the component as described in snapshot', () => {
    expect(fixture).toMatchSnapshot();
  });

  it('should open schedule dialog popup when onTimeSlotClick', () => {
    // given

    // when
    component.onTimeSlotClick({ hourIndex: 0, dayIndex: 0 });

    // then
    expect(scheduleDialogMock.open).toBeCalled();
  });

  it.skip('should open schedule dialog popup with valid params when onTimeSlotClick', (done) => {
    // given
    // let passedData = {};
    const indexes = { hourIndex: 0, dayIndex: 0 };
    const dialogData = component.getDialogData(indexes);
    scheduleDialogMock.open = jest.fn((comp, config) => {
      expect(config.data).toEqual(dialogData);
      done();
    });
    // when
    component.onTimeSlotClick(indexes);

    // then
    expect(scheduleDialogMock.open).toBeCalled();
  });

  it('should trigger student service update the class when schedule dialog is closed with data', (done) => {
    // given
    // (studentServiceMock.update as jest.Mock).mockResolvedValueOnce(Promise.resolve({ _id: 'updateclassid' }));

    // when
    component.onTimeSlotClick({ hourIndex: 0, dayIndex: 0 });
    let called = false;

    sub.add(
      (studentServiceMock.update = jest.fn(() => {
        called = true;
        done();
      })),
    );
    observableAfterClosed.next(mockedScheduleDialogData);

    // then
    expect(called).toBeTruthy();
  });

  it('should call student service getById when schedule dialog is closed with data', (done) => {
    // given
    let called = false;

    sub.add(
      (studentServiceMock.getById = jest.fn(() => {
        called = true;
        done();
      })),
    );

    // when
    component.onTimeSlotClick({ hourIndex: 0, dayIndex: 0 });
    observableAfterClosed.next(mockedScheduleDialogData);

    // then
    expect(called).toBeTruthy();
  });

  it('should call scheduleservice buildScheduleFromTimeslots when schedule dialog is closed with data', (done) => {
    // given
    let called = false;

    sub.add(
      (component.scheduleService.buildScheduleFromTimeslots = jest.fn(() => {
        called = true;
        done();
      })),
    );

    // when
    component.onTimeSlotClick({ hourIndex: 0, dayIndex: 0 });

    observableAfterClosed.next(mockedScheduleDialogData);

    // then
    expect(called).toBeTruthy();
  });

  function setup() {
    observableAfterClosed = new Subject();
    mockedScheduleDialogData = mockedClass.schedule;

    afterClosedMockFn = jest.fn().mockReturnValue(observableAfterClosed);

    scheduleDialogMock = {
      open: jest.fn().mockReturnValue({
        afterClosed: afterClosedMockFn,
      }),
    };

    const mockSchedule = [scheduleTestData];

    studentServiceMock = {
      // getById: jest.fn().mockReturnValue(Promise.resolve({})),
      update: jest.fn(),
      getById: jest.fn(),
    };

    TestBed.configureTestingModule({
      imports: [
        RouterModule.forRoot([]),
        TranslateModule.forRoot({
          loader: {
            provide: TranslateLoader,
            useClass: TranslateCustomLoader,
          },
        }),
      ],
      declarations: [StudentDetailsComponent, StudentDetailsHoursComponent],
      providers: [
        { provide: MatDialog, useValue: scheduleDialogMock },
        {
          provide: Router,
          useValue: {
            forRoot: jest.fn(),
          },
        },
        Platform,
        { provide: StudentService, useValue: studentServiceMock },
        ScheduleService,
        {
          provide: ActivatedRoute,
          useValue: {
            parent: {
              params: Observable.of({ idOrNew: '_new_' }),
            },
          },
        },
      ],
      schemas: [NO_ERRORS_SCHEMA],
    });

    fixture = TestBed.createComponent(StudentDetailsHoursComponent);
    component = fixture.componentInstance;
    component.schedule = mockSchedule;
    fixture.detectChanges();
  }
});
