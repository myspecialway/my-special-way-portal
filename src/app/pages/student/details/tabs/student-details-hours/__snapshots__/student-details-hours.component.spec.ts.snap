// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Student Details Hours Component with _new_ student path should render the component as described in snapshot 1`] = `
<app-student-details-hours
  dialog={[Function Object]}
  route={[Function Object]}
  scheduleService={[Function ScheduleService]}
  studentService={[Function StudentService]}
>
  <div
    class="msw-student-details-hours"
  >
    <app-schedule />
  </div>
</app-student-details-hours>
`;
